inc/msp/byte_vector.hpp

```diff
@@ -0,0 +1,263 @@
+        return rc;
+    }
+    
+    //unpack ByteVector data
+    bool unpack(ByteVector& val, size_t count = std::numeric_limits<size_t>::max()) {
+        if (count == std::numeric_limits<size_t>::max()) count = unpacking_remaining();        
+        if (!consume(count)) return false;
+        val.clear();
+        val.insert(val.end(),unpacking_iterator(),unpacking_iterator()+count);
+        return true;
+    }
+    
+    //unpack floating point types
+    //scale is the scale that already been applied to the value before encoding
+    //offset is the shift that was applied to the source value before encoding
+    template< typename encoding_T, typename T1, typename T2, 
+                typename std::enable_if<std::is_arithmetic<T1>::value, T1>::type* = nullptr,
+                typename std::enable_if<std::is_arithmetic<T2>::value, T2>::type* = nullptr >
+    bool unpack(T1& val, T2 scale, T2 offset = 0) {
+        bool rc = true;
+        encoding_T tmp = 0;
+        rc &= unpack(tmp);
+        val = tmp/scale;
+        val -= offset;
+        return rc;
+    }
+    
+    template<typename T, typename std::enable_if<std::is_base_of<Packable,T>::value, T>::type* = nullptr>
+    bool unpack(T obj) {
+        return obj.unpack_from(*this);
+    }
+    
+    //unpack value types other than string and ByteVector
+    template<class T>
+    bool unpack(value<T>& val) {
+        return val.set() = unpack(val()); 
+    }
+    
+    //unpack string value
+    
+    bool unpack(value<std::string>& val, size_t count = std::numeric_limits<size_t>::max()) {
+        return val.set() = unpack(val(),count); 
+    }
+    
+    bool unpack(value<ByteVector>& val, size_t count = std::numeric_limits<size_t>::max()) {
+        return val.set() = unpack(val(),count); 
+    }
+    
+    //unpack floating point value types
+    template< typename encoding_T, typename T1, typename T2, 
+                typename std::enable_if<std::is_arithmetic<T1>::value, T1>::type* = nullptr,
+                typename std::enable_if<std::is_arithmetic<T2>::value, T2>::type* = nullptr >
+    bool unpack(value<T1>& val, T2 scale = 1, T2 offset = 0) {
+        return val.set() = unpack<encoding_T>(val(),scale,offset);
+    }
+    
+    
+    //misc unpacking helper methods
+    std::size_t unpacking_offset() {
+        return offset;
+    }
+    
+    std::vector<uint8_t>::iterator unpacking_iterator() {
+        return this->begin()+offset;
+    }
+    
+    bool consume(std::size_t count) {
+        if (count < -offset) return false;
+        if (count > this->size() - offset) return false;
+        offset += count;
+        return true;
+    }
+    
+    std::size_t unpacking_remaining() {
+        return this->size() - offset;
+    }
+
+protected:
+    std::size_t offset;
+
+};
+
+
+
+struct Packable {
+    virtual bool pack_into(ByteVector &data) const = 0;
+    virtual bool unpack_from(ByteVector &data) = 0;
+};
+
+typedef std::shared_ptr<ByteVector> ByteVector_ptr;
```

Comment:
Use `ByteVectorPtr` for this.

Reply:
I actually just meant that you rename the type, i.e.:
`typedef std::shared_ptr<ByteVector> ByteVectorPtr;`
because I prefer CamelCase type names.